### PR TITLE
ci-operator: stop allowing users to set artifact dir

### PR DESCRIPTION
**Pull request by stevekuznetsov**
No longer are users able to set the explicit directory for:
 - `ci-operator` outputs
 - simple Pod tests
 - multi-stage test steps

`ci-operator` itself will read the `${ARTIFACTS}` value to determine
where to put its own artifacts (like the output jUnit files). This is in
effect what already happens, as this is the fallback to when the flag is
not passed, and we do not pass it in generated jobs. The flag will
remain in a deprecated manner so as to not break users.

All test steps *except* for `Template` tests are already using the
test-infra pod-utils to upload, and therefore are not allowed to set a
non-standard directory, anyway. Nothing changes for these steps. It is
now not possible to opt out of this by telling `ci-operator` not to use
pod-utils.

We continue to use the deprecated artifact fetching sidecar code for
Templates, but only for templates.

This will cause a breaking change to prowgen as our previous automation
explicitly wrote our defaulted value into all configs even if they never
set it.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @alvaroaleman @petr-muller @bbguimaraes

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Part of DPTP-1668

File: pkg/steps/artifacts.go [hunk 1]
```diff
@@ -366,27 +367,6 @@ func addPodUtils(pod *coreapi.Pod, artifactDir string, decorationConfig *prowv1.
 	return nil
 }
 
-func addArtifacts(pod *coreapi.Pod, artifactDir string) {
```
Comment (stevekuznetsov) [inline pkg/steps/artifacts.go:369]: This is suspicious but I think templates don't actually use this, they use `addArtifactsToPod` ...

**Comment by stevekuznetsov** [comment]
Hmm I think I can split this up

**Comment by stevekuznetsov** [comment]
OK, so it theoretically should be possible to split this into steps:

 1. remove the feature from basic container jobs
 2. remove it from step registry stuff
 3. remove it from the ci-operator top-level

I'll try to do that - there are areas where we infer 1&2 from 3 randomly so it's a bit messy but probably doable.

**Comment by stevekuznetsov** [comment]
@alvaroaleman @petr-muller updated this to be minimal now that the rest of the changes have come in. Notably, this *removes* the `--artifact-dir` setting from `ci-operator` itself. I think this is a useful thing to do as:

1. we never set it explicitly
2. we only care about using it when Prow's running us

In `openshift/release` today, we have this:

```
$ git grep -h artifact-dir | head -n 1
        - --artifact-dir=$(ARTIFACTS)
$ git grep artifact-dir | wc -l
563
$ git grep artifact-dir | grep -v ARTIFACTS | wc -l
0
```

So there's 563 places where we set it to what would otherwise be the default value anyway, so there's zero utility to this flag. In addition, having to pass this directory around _everywhere_ in our codebase ended up being a bad cargo-cult ... half the places didn't even end up using it, and in some cases like multi-stage it looks like we accidentally used this instead of the per-step one ...

File: images/entrypoint-wrapper/Dockerfile [hunk 2]
```diff
@@ -3,3 +3,4 @@ LABEL maintainer="bbarcaro@redhat.com"
 
 ADD entrypoint-wrapper /usr/bin/entrypoint-wrapper
 ENTRYPOINT ["/usr/bin/entrypoint-wrapper"]
+
```
Comment (alvaroaleman) [inline images/entrypoint-wrapper/Dockerfile:6]: ?

File: pkg/util/env.go [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package util
+
+import "os"
+
+// prowArtifactsEnv is the directory Prow wants us to put artifacts into for upload
+const prowArtifactsEnv string = "ARTIFACTS"
```
Comment (alvaroaleman) [inline pkg/util/env.go:6]: IMHO this is not util, this is `api`

File: images/entrypoint-wrapper/Dockerfile (hunk 2 above)
Comment (stevekuznetsov) [reply images/entrypoint-wrapper/Dockerfile:6]: fat fingers

File: cmd/ci-operator/main.go [hunk 4]
```diff
@@ -186,6 +183,9 @@ func main() {
 		if err := flag.CommandLine.Set("v", "10"); err != nil {
 			logrus.WithError(err).Error("Failed to set flag -v=10")
 		}
+		logrus.SetLevel(logrus.TraceLevel)
```
Comment (alvaroaleman) [inline cmd/ci-operator/main.go:186]: ?

File: cmd/ci-operator/main.go [hunk 5]
```diff
@@ -345,7 +345,7 @@ func bindOptions(flag *flag.FlagSet) *options {
 	flag.BoolVar(&opt.promote, "promote", false, "When all other targets complete, publish the set of images built by this job into the release configuration.")
 
 	// output control
-	flag.StringVar(&opt.artifactDir, "artifact-dir", "", "If set grab artifacts from test and template jobs. Defaults to $ARTIFACTS if set.")
+	flag.StringVar(&opt.artifactDir, "artifact-dir", "", "DEPRECATED.")
```
Comment (alvaroaleman) [inline cmd/ci-operator/main.go:348]: Please also add the info that it doesn't do anything

File: images/entrypoint-wrapper/Dockerfile [hunk 6]
```diff
@@ -2,4 +2,4 @@ FROM centos:8
 LABEL maintainer="bbarcaro@redhat.com"
 
 ADD entrypoint-wrapper /usr/bin/entrypoint-wrapper
-ENTRYPOINT ["/usr/bin/entrypoint-wrapper"]
+ENTRYPOINT ["/usr/bin/entrypoint-wrapper"]
```
Comment (alvaroaleman) [inline images/entrypoint-wrapper/Dockerfile:5]: ?

File: cmd/ci-operator/main.go (hunk 4 above)
Comment (stevekuznetsov) [reply cmd/ci-operator/main.go:186]: This was part of my commit to add logging to the artifacts thing. The default log library we use has no concept of levels, and we use logrus anyway, so I just added logging to it in Trace level, and setting ci-operator verbose mode opens that up.

File: images/entrypoint-wrapper/Dockerfile (hunk 6 above)
Comment (stevekuznetsov) [reply images/entrypoint-wrapper/Dockerfile:5]: sigh

**Comment by stevekuznetsov** [comment]
@alvaroaleman updated

**Comment by stevekuznetsov** [comment]
Seeing a lot of this lately

```
 error: error creating buildah builder: Error writing blob: error storing blob to file "/var/tmp/storage350157992/5": unexpected EOF 
```

/retest

**Comment by stevekuznetsov** [comment]
Rebased and ready @alvaroaleman @petr-muller

File: cmd/ci-operator/main.go (hunk 4 above)
Comment (alvaroaleman) [reply cmd/ci-operator/main.go:186]: I think my larger point was that almost everywhere we run logrus with Debug level by default and reserve trace level logs for things that you need to opt in to get them, I think it is valuable to be consistent with that

Comment (petr-muller) [reply cmd/ci-operator/main.go:186]: Well this *is* opt-in, it's under verbose so I guess this is fine?

**Review by petr-muller — APPROVED** [review]
LGTM

/hold
Holding to allow resolving Alvaro's comment

File: cmd/ci-operator/main.go (hunk 4 above)
Comment (stevekuznetsov) [reply cmd/ci-operator/main.go:186]: Yeah you need to `--verbose` the `ci-operator` binary to get this, which you'd only be doing in dev.

**Comment by stevekuznetsov** [comment]
/hold cancel

You need to opt in via flag

**Pull request merged**